File: kotlin/0930-binary-subarrays-with-sum.kt
class Solution {
    fun numSubarraysWithSum(nums: IntArray, goal: Int): Int {

        fun helper(x: Int): Int {
            if (x < 0) return 0

            var res = 0
            var l = 0
            var cur = 0
            for (r in 0 until nums.size) {
                cur += nums[r]

                while (cur > x)
                    cur -= nums[l++]

                res += (r - l + 1)
            }

            return res
        }

        return helper(goal) - helper(goal - 1)
    }
}

File: python/0930-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:

        def helper(x):
            if x < 0: return 0

            res = 0
            l, cur = 0, 0
            for r in range(len(nums)):
                cur += nums[r]
                while cur > x:
                    cur -= nums[l]
                    l += 1
                res += (r - l + 1)
            return res
        
        return helper(goal) - helper(goal - 1)
